rbd: replace trash delay option, add rbd trash purge command

thmour opened the pull request:
There was a discussion about the trash delay option (that is in seconds) to be replaced with something else. @dvanders proposes a `/bin/date` style input of the form of `2017-12-30`, `now`, `1 week ago`, etc. It is a human-readable text and it can't be confused with something like `60*24*7`. This pull request implements it, by calling `/bin/date` safely ([exec](https://linux.die.net/man/3/exec)*).

There is also a need for a batch removal of images in the trash that have expired. It can be used on Openstack as a cron job. The name of the command is "purge" and deletes every image with expired deferment end time from the defined pool. There is also an option `--older-than` implemented to remove images with deferment end time older than a given date.

This pull request is still a work in progress and we would like to get some feedback possibly from @dillaman and @rjfd  


Thanks

dillaman commented on src/include/o_execv.h:
```diff
@@ -0,0 +1,36 @@
+#include <stdio.h>
```
I believe there are existing wrappers around ```execv``` like ```common/SubProcess.h```

thmour replied:
thanks, I was searching for them

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -260,7 +260,7 @@ CEPH_RBD_API int rbd_rename(rados_ioctx_t src_io_ctx, const char *srcname,
                             const char *destname);
 
 CEPH_RBD_API int rbd_trash_move(rados_ioctx_t io, const char *name,
-                                uint64_t delay);
```
Since the API is released, we cannot create a breaking change. The human-readable parsing can probably be limited to the rbd CLI since any API users will have no issue with supplying a derived time in seconds.

dillaman replied:
I still think this comment is valid -- from an API perspective, it doesn't make much sense to have a negative delay. If some uses the rbd CLI to specify a time in the past, it should be warned/prohibited.

thmour replied:
In our use cases we need to set the expiration times to the past for test purposes, debugging and averting mistakes that cost space, we don't like to have unexpected things. In the end, a date is passed to the trashspec, so why do we have to use an offset as an argument instead of just passing the desired date as an absolute value? Because now to set a protected until date for an image, I have to get the date, subtract from now, pass the offset to the function so that it adds the offset from now to store the date that I initially had. With having the "uint64_t delay" as "time_t protected_until", I could just put "time(NULL)+5" in C or "int(time.time())+5" in Python2 and have the same effect. This positive offset has really no flexibility and serves a narrow variety of use cases.

dillaman replied:
The issue is that we cannot change a released API -- so now such a change would require a ```rbd_trash_move2(..., bool expire_time_valid, time_t expire_time)```-like method. The reason that an offset delay was chosen was it solved the issue of specifying a time and also indicating whether or not a "protection period" was required (i.e. any value greater than zero).

dillaman commented on src/include/utime.h:
```diff
@@ -322,6 +325,23 @@ class utime_t {
         bdt.tm_hour, bdt.tm_min, bdt.tm_sec);
   }
 
+  static int invoke_date(const std::string date_str, utime_t *result) {
```
Nit: pass ```date_str``` by const reference

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -36,8 +36,8 @@ void get_move_arguments(po::options_description *positional,
   options->add_options()
-    (at::DELAY.c_str(), po::value<uint64_t>(),
-     "time delay in seconds until effectively remove the image");
+    (at::EXPIRES_IN.c_str(), po::value<std::string>()->default_value("1 week"),
```
Nit: why not default to "now" since "1 week" otherwise seems arbitrary?

thmour replied:
It is still in discussion, we wanted some safe date in case someone mistakenly trashes an image and a trash purge cron job runs after a second and deletes it (and then the image can't be restored). It can be changed to a "day", or "now" if it makes more sense.

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -151,6 +150,8 @@ std::string delete_status(time_t deferment_end_time) {
   if (now < deferment_end_time) {
     ss << "protected until " << time_str;
+  } else {
+    ss << "expired at      " << time_str;
```
Nit: lots of extra spaces

thmour replied:
it results in clean readable output, imagine many "expired at" and "protected until" one after another
```
[11:06][root@xxx (ceph_dev:ceph/build*10) build]# bin/rbd trash list --pool mypool --long
ID           NAME SOURCE DELETED_AT               STATUS                                   
10d174b0dc51 bar  USER   Thu Oct 12 10:51:23 2017 expired at      Fri Oct 13 10:51:23 2017 
10d72ae8944a faz  USER   Wed Oct 18 11:06:08 2017 protected until Fri Oct 20 11:06:08 2017 
10eb2ae8944a foo  USER   Mon Oct 16 11:25:10 2017 expired at      Mon Oct 16 11:25:10 2017 
11032ae8944a baz  USER   Wed Oct 18 11:01:02 2017 protected until Thu Oct 19 11:01:02 2017 
11262ae8944a laz  USER   Wed Oct 18 11:05:19 2017 expired at      Tue Oct 17 11:05:19 2017 
```

dillaman replied:
... true, but if using JSON/XML formatted output, the spacing is unnecessary.

thmour replied:
I am sorry, thanks again!

dillaman replied:
No worries -- thx

thmour commented:
I now replaced the delay option only in the cli, I also added a trash_move2 function to the c/cpp api instead of breaking the current one. Is it better now?

dillaman commented:
It still seems odd to me to have an API take an English-parseable parameter to specify the delay -- plus that parameter is then fed to a external application to actually parse it. That's why I think I would still vote for leaving the ```trash_move``` API as it is but then incorporate the user-friendly parsing within the rbd CLI (which passes the binary delay to the API).

thmour commented:
You are right. I should move this functionality to the cli scope. I just used it in the internal.cc to be able to refer to the past (because the delay argument is unsigned), so I can debug the purge command, but letting delay accept negative values could result in the same flexibility as a date string.

thmour commented:
From the api i just changed the delay from uint64_t to int64_t, I use now the /bin/date only on the rbd cli

thmour commented:
I left the api intact, I focus on the cli now. There is a feature needed called threshold on rbd purge that tries to bring the pool to a certain data usage by removing the oldest trashed images based on the deferment end time. When I use `rados.get_pool_stats` after the `rbd.trash_remove` I don't see the updated pool data usage. Is there any better way than this `do...while` to do this task?

dillaman commented:
@thmour The usage stats are periodically updated over long periods (>30 seconds) so you wouldn't be able to rely on them within a tight loop.

thmour commented:
@dillaman I changed the algorithm to get the disk usage of each image and sum them to reach the threshold instead of deleting images one by one and polling the usage until it reaches the threshold. It seems to have a satisfactory performance (deleted 100GB images with ~50GB disk usage, took a couple of seconds per image). Is it ready to merged? Does the purge command need some kind of test units? I ask because I see no tests for the cli commands, other than the "help.t" for the command output, because the other commands are tested through their equivalent APIs.

dillaman commented:
@thmour CLI commands don't currently get unit tests but they do have integration tests [1]

[1] https://github.com/ceph/ceph/blob/master/qa/workunits/rbd/cli_generic.sh#L386

thmour commented:
@dillaman Added a integration test for the trash purge. Anything else?

dillaman commented:
@thmour Can you run "git rebase origin/master" and "git push origin thmour_rbdtrash -f" to remove that merge commit?

dillaman commented on src/librbd/internal.h:
```diff
@@ -107,7 +107,7 @@ namespace librbd {
              const std::string &image_id, ProgressContext& prog_ctx,
              bool force=false, bool from_trash_remove=false);
-
+ 
```
Nit: added white space?

dillaman commented on src/tools/rbd/ArgumentTypes.h:
```diff
@@ -81,7 +81,9 @@ static const std::string PRETTY_FORMAT("pretty-format");
 
-static const std::string DELAY("delay");
+static const std::string EXPIRES_AT("expires-at");
+static const std::string OLDER_THAN("older-than");
```
Nit: perhaps for clarity the option should be ```expired-before``` since it's based on the expiration time and not the deletion time?

dillaman commented on src/tools/rbd/ArgumentTypes.h:
```diff
@@ -81,7 +81,9 @@ static const std::string PRETTY_FORMAT("pretty-format");
 static const std::string NO_ERROR("no-error");
 
-static const std::string DELAY("delay");
+static const std::string EXPIRES_AT("expires-at");
```
Nit: since these arguments are really only used by ```Trash.cc```, we can move them all there while making these changes.

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -31,13 +31,14 @@ namespace trash {
 namespace at = argument_types;
 namespace po = boost::program_options;
 
+typedef librbd::trash_image_info_t trash_info_t;
```
Nit: why rename the type?

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -60,20 +61,30 @@ int execute_move(const po::variables_map &vm) {
   librbd::RBD rbd;
-  r = rbd.trash_move(io_ctx, image_name.c_str(), delay);
+  
+  utime_t t_exp, now = ceph_clock_now();
```
Nit: declare variables separately for clarity (especially so w/ assignments).

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -60,20 +61,30 @@ int execute_move(const po::variables_map &vm) {
+  }
+  int64_t dt = t_exp.sec() - now.sec();
+  if(dt < 0) {
+    std::cerr << "rbd: can not use a date in the past as an expiration date" << std::endl;
```
Nit: ```cannot``` vs ```can not```

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -60,20 +61,30 @@ int execute_move(const po::variables_map &vm) {
-  r = rbd.trash_move(io_ctx, image_name.c_str(), delay);
+  
+  utime_t t_exp, now = ceph_clock_now();
+  r = utime_t::invoke_date(expires_at, &t_exp);
```
Nit: it would be nice to only invoke ```date``` if an expiration time was specified

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -60,20 +61,30 @@ int execute_move(const po::variables_map &vm) {
+  utime_t t_exp, now = ceph_clock_now();
+  r = utime_t::invoke_date(expires_at, &t_exp);
+  if (r < 0) {
+    return r;
```
Nit: add error message since this spawning of ```date``` could be problematic.

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
+      (at::OLDER_THAN.c_str(), po::value<std::string>()->value_name("date"), 
+       "purges images that expired before the given date");
+  options->add_options()
+      (at::THRESHOLD.c_str(), po::value<double>(), "purges images until the current pool data usage is reduced to X%, value range: 0.0-1.0");
```
Nit: watch line length -- try to keep under 80 characters

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
+  
+  std::vector<trash_info_t> trash_entries;
+  r = rbd.trash_list(io_ctx, trash_entries);
+  if (r < 0) { return r; }
```
Nit: use multiple lines

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
+      if(bytes_to_free >= bytes_threshold) break;
+    }
+  }
+  else {
```
Nit: move ```else {``` back up to previous line

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
+    rados.get_pool_stats(vec, stats);
+
+    uint64_t bytes_to_free = 0;
+    auto bytes_threshold = (uint64_t)(stats[pool_name].num_bytes * (1 - threshold));
```
So the threshold is based on the used space of the pool and not the max available space? It seems if you are trying to ensure a pool has X% free space available, this calculation should be based upon max available. This information is available by sending a ```mon_command``` of ```{'prefix': 'df', 'format': 'json'}``` via librados and using json_spirit to find the pool and parsing out ```max_avail```.

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
+      if(r < 0) continue;
+      
+      uint64_t img_size; curr_img.size(&img_size);
+      r = curr_img.diff_iterate2(nullptr, 0, img_size, false, false, 
```
If you change that second ```false``` to ```true``` you will enable fast-diff space calculation which will be much faster at the expense of rounding to 4MB object sizes.

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -300,6 +317,113 @@ int execute_list(const po::variables_map &vm) {
+    if (vm.find(at::OLDER_THAN) != vm.end()) { 
+      utime_t new_time;
+      r = utime_t::invoke_date(vm[at::OLDER_THAN].as<std::string>(), &new_time);
+      if (r < 0) return r;
```
Nit: add error message when ```date``` fails

thmour commented:
@dillaman A big thanks for your comments!
I added, filled and trashed 4x64M images in my vstart cluster, they filled the 'test' pool by 2.81%.
Then I executed:
```bash
$ bin/rbd trash purge test --threshold 0.015 #1.5%
Removing images: 100% complete...done.
$ bin/ceph df
GLOBAL:
    SIZE      AVAIL     RAW USED     %RAW USED 
    30.2G     26.4G        3.77G         12.50 
POOLS:
    NAME                  ID     USED      %USED     MAX AVAIL     OBJECTS 
    cephfs_data_a         1          0         0         8.70G           0 
    cephfs_metadata_a     2      2.19K         0         8.70G          21 
    test                  3       128M      1.42         8.70G          39 
```
Works as intended. As the threshold test can't be hardcoded I added to the test the trivial 0 and 1 values.

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -320,6 +345,148 @@ int execute_list(const po::variables_map &vm) {
+   
+
+    json_spirit::mValue json;
+    json_spirit::read_or_throw(outbl.to_str(), json);
```
Nit: would be nicer to not throw and instead log an error and return an error code if you couldn't parse the result

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -320,6 +345,148 @@ int execute_list(const po::variables_map &vm) {
+  librbd::RBD rbd;
+  
+  std::vector<librbd::trash_image_info_t> trash_entries;
+  r = rbd.trash_list(io_ctx, trash_entries);
```
I was thinking that ```purge``` should probably only handle images where the source == ```USER``` to avoid deleting RBD-internally created trash entries (i.e. rbd-mirror deferred deletions and/or migrating images).

thmour replied:
done

dillaman commented on src/tools/rbd/action/Trash.cc:
```diff
@@ -320,6 +345,159 @@ int execute_list(const po::variables_map &vm) {
+      if(bytes_to_free >= bytes_threshold) break;
+    }
+  } else {
+    time_t expire_ts = ceph_clock_gettime();
```
Nit: can you rebase your PR since this method was renamed to ```clock_gettime```?

dillaman commented:
@thmour Can you also update the ```cli_generic.sh``` test since it's now failing due to the removal of the ```--delay``` optional [1]

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2018-01-02_14:09:46-rbd-wip-jd-testing-distro-basic-smithi/2016487/teuthology.log

dillaman commented:
retest this please
